query.go

```diff
@@ -3,317 +3,362 @@ package dht
-	proc := process.WithParent(process.Background())
-	ctx := ctxproc.OnClosingContext(proc)
-	peersToQuery := queue.NewChanQueue(ctx, queue.NewXORDistancePQ(string(q.key)))
-	r := &dhtQueryRunner{
-		query:          q,
-		peersRemaining: todoctr.NewSyncCounter(),
-		peersSeen:      peer.NewSet(),
-		peersQueried:   peer.NewSet(),
-		rateLimit:      make(chan struct{}, q.concurrency),
-		peersToQuery:   peersToQuery,
-		proc:           proc,
-	}
-	dq, err := newDialQueue(&dqParams{
-		ctx:    ctx,
-		target: q.key,
-		in:     peersToQuery,
-		dialFn: r.dialPeer,
-		config: dqDefaultConfig(),
-	})
-	if err != nil {
-		panic(err)
+		for _, p := range peersToQuery[:numQuery] {
+			go func(p peer.ID) {
+				queryResCh <- q.queryPeer(q.ctx, p)
+			}(p)
+		}
+
+	loop:
+		for {
+			select {
+			case res := <-queryResCh:
+				foundCloser = foundCloser || res.foundCloserPeer
+				resultsReceived++
+				if resultsReceived == numQuery {
+					break loop
+				}
+			case <-q.ctx.Done():
+				return
+			}
+		}
 	}
-	r.peersDialed = dq
-	return r
 }
 
-func (r *dhtQueryRunner) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, error) {
-	// Make sure to clean everything up when we return from this function.
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+func simpleQuery(q *query) {
+	/*
+		start with K closest peers (some queried already some not)
+		take best alpha (sorted by some metric)
+		query those alpha
+		   - if a query fails then take the next one
+		once they complete:
+			if the alpha requests did not add any new peers to top K, repeat with unqueried top K
+			else repeat
+	*/
+
+	var lastPeers []peer.ID
+	for {
+		peersToQuery := q.localPeers.KUnqueried()
 
-	r.log = logger
-	r.runCtx = ctx
+		if len(peersToQuery) == 0 {
+			return
+		}
 
-	// setup concurrency rate limiting
-	for i := 0; i < r.query.concurrency; i++ {
-		r.rateLimit <- struct{}{}
-	}
+		numQuery := q.dht.alpha
+		if lastPeers != nil && peerSlicesEqual(lastPeers, peersToQuery) {
+			numQuery = len(peersToQuery)
+		} else if pqLen := len(peersToQuery); pqLen < numQuery {
+			numQuery = pqLen
+		}
 
-	// add all the peers we got first.
-	for _, p := range peers {
-		r.addPeerToQuery(p)
+		peersToQueryCh := make(chan peer.ID, numQuery)
+		for _, p := range peersToQuery[:numQuery] {
+			peersToQueryCh <- p
+		}
+		queryResCh := make(chan *queryResult, numQuery)
+		queriesSucceeded, queriesSent := 0, numQuery
+
+	dialPeers:
+		for {
+			select {
+			case p := <-peersToQueryCh:
+				go func() {
+					queryResCh <- q.queryPeer(q.ctx, p)
+				}()
+			case res := <-queryResCh:
+				if res.success {
+					queriesSucceeded++
+					if queriesSucceeded == numQuery {
+						break dialPeers
+					}
+				} else {
+					queriesSent++
+					if queriesSent >= len(peersToQuery) {
+						break dialPeers
+					}
+					peersToQueryCh <- peersToQuery[queriesSent]
+				}
+			case <-q.ctx.Done():
+				return
+			}
+		}
 	}
+}
+
+func boundedDialQuery(q *query) {
```

Comment:
This function isn't getting used anywhere.

Reply:
TLDR: We should either delete the function or expose it as an option before we merge this PR

Ya... the idea was that the actual way in which we concurrently send out queries based on our concurrency parameter `alpha` is AFAIK not strictly specified by Kademlia. This means there are potentially a few viable candidates out there to play with and gain different performance profiles. The one we have now is "alpha at a time... wait... alpha again - stop when we stop getting new peers in our top K". This is the simplest option, but not necessarily the most performant.

Unfortunately, I haven't yet had time + priority to do testing/profiling and figure out how differently these different functions (including https://github.com/libp2p/go-libp2p-kad-dht/pull/473#discussion_r388925013) perform.

Certainly before we merge this we should either delete this code, or make these query functions options (either as an enum or as a function).